File: Go/greedy.go
package main

import (
	"os"
    "bufio"
    "bytes"
    "io"
    "fmt"
	"math"
	"strings"
	"strconv"
	)

type Activity struct {
	id string
	lat float64
	lon float64
}

type Resource struct {
	id string
	lat float64
	lon float64
}

type Allocation struct {
	rid string
	aid string
	dist float64
}

type SchemaData struct {
	activity map[string]Activity 
	resource map[string]Resource
	allocation []Allocation
}

var earthRadiusM float64 = 6367450
var convert2Rad float64 = math.Pi / 180.0
var convert2Deg float64 = 180.0 / math.Pi
var seconds_per_metre float64 = 0.0559234073

func readLines(path string) (lines []string, err error) {
    var (
        file *os.File
        part []byte
        prefix bool
    )
    if file, err = os.Open(path); err != nil {
        return
    }
    defer file.Close()

    reader := bufio.NewReader(file)
    buffer := bytes.NewBuffer(make([]byte, 0))
    for {
        if part, prefix, err = reader.ReadLine(); err != nil {
            break
        }
        buffer.Write(part)
        if !prefix {
            lines = append(lines, buffer.String())
            buffer.Reset()
        }
    }
    if err == io.EOF {
        err = nil
    }
    return
}

func distanceBetweenPointsLatLong(lat1 float64, lon1 float64, lat2 float64, lon2 float64) float64 {
		dStartLatInRad := lat1 * convert2Rad
		dStartLongInRad := lon1 * convert2Rad
		dEndLatInRad := lat2 * convert2Rad
		dEndLongInRad := lon2 * convert2Rad
		dLongitude := dEndLongInRad - dStartLongInRad
		dLatitude := dEndLatInRad - dStartLatInRad
		dSinHalfLatitude := math.Sin(dLatitude * 0.5)
		dSinHalfLongitude := math.Sin(dLongitude * 0.5)
		a := dSinHalfLatitude * dSinHalfLatitude + math.Cos(dStartLatInRad) * math.Cos(dEndLatInRad) * dSinHalfLongitude * dSinHalfLongitude
		c := math.Atan2(math.Sqrt(a), math.Sqrt(1.0 - a))
		return earthRadiusM * (c + c)
}

func scheduleResources(sd *SchemaData) {
	for xres, res := range sd.resource {
		for c := 0; c < 5; c++ {
			lowest := math.Inf(0)
			var lowestid = ""
			for xact, act := range sd.activity {
				dist := distanceBetweenPointsLatLong(res.lat, res.lon, act.lat, act.lon)
				if dist < lowest {
					lowest = dist
					lowestid = xact
				}
			}
			a := Allocation { rid:xres, aid:lowestid, dist:lowest}
			sd.allocation = append (sd.allocation, a)
			delete(sd.activity, lowestid)
		}
	}
}

func build(lines []string) SchemaData {
	sd := SchemaData {}
	sd.activity = map[string]Activity {}
	sd.resource = map[string]Resource {}
	for _, line := range lines {
		split := strings.Split(line, ",")
		lat, _ := strconv.ParseFloat(split[1], 64)
		lon, _ := strconv.ParseFloat(split[2], 64)
  		switch len(split) {
  		case 3:
  			r := Resource { id:split[0], lat:lat, lon:lon }
  			sd.resource[r.id] = r
  		case 4:
  			a := Activity { id:split[0], lat:lat, lon:lon }
  			sd.activity[a.id] = a
  		}
    }
    return sd
}

func main() {
	lines, _ := readLines("/Users/daryl/Development/Projects/FunctionalComparison/Data/DataSPIF.csv")
    for i := 0; i < 1000; i++ {
    	sd := build(lines)
	    scheduleResources(&sd)
		sum := 0.0
		for _, alloc := range sd.allocation {
			sum += alloc.dist
		}
		fmt.Println(sum)	
	}
}

File: Julia/greedy.jl
struct Activity
	id::String
	lat::Float64
	lon::Float64
    Activity(id,lat,lng) = new(id,lat,lng)
end

struct Resource
	id::String
	lat::Float64
	lon::Float64
    Resource(id,lat,lng) = new(id,lat,lng)
end

struct Allocation
	rid::String
	aid::String
	dist::Float64
    Allocation(rid,aid,dist) = new(rid,aid,dist)
end

struct SchemaData
    activity::Array{Activity}
    resource::Array{Resource}
    allocation::Array{Allocation}
    SchemaData(activity,resource,allocation) = new(activity,resource,allocation)
end

activities = Activity[]
resources = Resource[]
open("../Data/DataSPIF.csv") do f
    lines = readlines(f)
    for line in lines
        items = split(line, ",")
        l = length(items)
        if l == 3
            r = Resource(items[1],parse(Float64,items[2]),parse(Float64,items[3]))
            push!(resources,r)
            #println(r)
        else
            a = Activity(items[1],parse(Float64,items[2]),parse(Float64,items[3]))
            push!(activities,a)
            #println(a)
        end
    end
end

function distanceBetweenPointsLatLong(lat1, lon1, lat2, lon2)
    const earthRadiusM::Float64 = 6367450.0
    const convert2Rad::Float64 = pi / 180.0
    const convert2Deg::Float64 = 180.0 / pi
    const seconds_per_metre::Float64 = 0.0559234073
    dStartLatInRad = lat1 * convert2Rad
    dStartLongInRad = lon1 * convert2Rad
    dEndLatInRad = lat2 * convert2Rad
    dEndLongInRad = lon2 * convert2Rad
    dLongitude = dEndLongInRad - dStartLongInRad
    dLatitude = dEndLatInRad - dStartLatInRad
    dSinHalfLatitude =  sin(dLatitude * 0.5)
    dSinHalfLongitude = sin(dLongitude * 0.5)
    a = dSinHalfLatitude * dSinHalfLatitude + cos(dStartLatInRad) * cos(dEndLatInRad) * dSinHalfLongitude * dSinHalfLongitude
    c = atan2(sqrt(a), sqrt(1.0 - a))
    earthRadiusM * (c + c)
end

function scheduleResources(sd)
    for res in sd.resource 
        for c in 1:50
            lowest = typemax(Float64)
            lowestact = Activity("",0.0,0.0)
            lowestidx = 0
            for (i,act) in enumerate(sd.activity)
                dist = distanceBetweenPointsLatLong(res.lat, res.lon, act.lat, act.lon)
                if dist < lowest
                    lowest = dist
                    lowestact = act 
                    lowestidx = i
                end
            end
            push!(sd.allocation,Allocation(res.id, lowestact.id, lowest))
            deleteat!(sd.activity, lowestidx)
        end
    end
end

function run()
    earthRadiusM = 6367450.0
    convert2Rad = pi / 180.0
    convert2Deg = 180.0 / pi
    seconds_per_metre = 0.0559234073

    for i in 1:100
        sd = SchemaData(copy(activities),copy(resources),Allocation[])
        scheduleResources(sd)
        f = map(x -> x.dist,sd.allocation)
        println(string(i) * ":" * string(sum(f)))
    end
end

run()